### PR TITLE
Feat/better testing for deals awaiting seals

### PR DESCRIPTION
**Pull request by acruikshank**
This PR is a refactor that ballooned out of #2693 into enough work for a new PR.

### Problem

dealsAwaitingSeals is burried within protocol/storage/miner. I found its logic confusing and under-tested. 

### Solution

This PR streamlines the logic, moves it to its own file with a separate file for tests, and adds testing for its features.

### CODE REVIEW

File: protocol/storage/deals_awaiting_seal_test.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+	wantMessage := "boom"
+
+	t.Run("add before success", func(t *testing.T) {
+		dealsAwaitingSeal := newDealsAwaitingSeal()
+		gotCids := []cid.Cid{}
+		dealsAwaitingSeal.onSuccess = func(dealCid cid.Cid, sector *sectorbuilder.SealedSectorMetadata) {
+			assert.Equal(t, sector, wantSector)
+			gotCids = append(gotCids, dealCid)
+		}
```
Comment (anorth) [inline protocol/storage/deals_awaiting_seal_test.go]: Set `onFail` too and check that it's never called. Ditto for the other tests: check both successes and failures.

File: protocol/storage/deals_awaiting_seal.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	ErrorMessage     string
+}
+
+// dealsAwaitingSeal is a container for keeping track of which sectors have
+// pieces from which deals. We need it to accommodate a race condition where
+// a sector commit message is added to chain before we can add the sector/deal
+// book-keeping. It effectively caches success and failure results for sectors
+// for tardy add() calls.
+type dealsAwaitingSeal struct {
+	l sync.Mutex
```
Comment (frrist) [inline protocol/storage/deals_awaiting_seal.go]: nit: can we call this dealsMu (I like [hats](https://dmitri.shuralyov.com/idiomatic-go/entries/2))

File: protocol/storage/deals_awaiting_seal.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	// Don't keep references to sectors around forever. Assume that at most
+	// one success-before-add call will happen (eg, in a test). Sector sealing
+	// outside of tests is so slow that it shouldn't happen in practice.
+	// So now that it has happened once, clean it up. If we wanted to keep
+	// the state around for longer for some reason we need to limit how many
+	// sectors we hang onto, eg keep a fixed-length slice of successes
+	// and failures and shift the oldest off and the newest on.
+	delete(dealsAwaitingSeal.SealedSectors, sectorID)
```
Comment (frrist) [inline protocol/storage/deals_awaiting_seal.go]: In the same vein as @anorth's comment -- This method is called `add` yet will sometimes perform a delete operation, maybe we could call this `addAndCleanup` or `modify`?

File: protocol/storage/deals_awaiting_seal.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+func init() {
+	cbor.RegisterCborType(dealsAwaitingSeal{})
+}
+
+// sectorInfo combines sector Metadata from rust proofs with go-filecoin specific data
+type sectorInfo struct {
+	Metadata         *sectorbuilder.SealedSectorMetadata
+	CommitMessageCid *cid.Cid
```
Comment (frrist) [inline protocol/storage/deals_awaiting_seal.go]: nit: does this need to be a pointer? We can open ourselves up to some [gotchas](https://github.com/filecoin-project/go-filecoin/blob/master/CONTRIBUTING.md#gotchas) with this. My preference would be to use a value and return `cid.Undef` inplace of `nil`.

File: protocol/storage/deals_awaiting_seal_test.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	"github.com/filecoin-project/go-filecoin/proofs/sectorbuilder"
+	tf "github.com/filecoin-project/go-filecoin/testhelpers/testflags"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/ipfs/go-cid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDealsAwaitingSealAdd(t *testing.T) {
+	tf.UnitTest(t)
```
Comment (frrist) [inline protocol/storage/deals_awaiting_seal_test.go]: :pray:

**Comment by acruikshank** [comment]
This didn't get merged before release and then got left behind. It's in a nearly irreconcilable state now. I gave up and cherry picked the commits and responded to remaining feedback here: #2791

**Pull request closed without merging**